File: swift/Sources/benchmarks/AStarPathfinder.swift
import Foundation

final class AStarPathfinder: BenchmarkProtocol {
  private struct Node: Comparable {
    let x: Int
    let y: Int
    let fScore: Int

    static func < (lhs: Node, rhs: Node) -> Bool {

      if lhs.fScore != rhs.fScore {
        return lhs.fScore < rhs.fScore
      }
      if lhs.y != rhs.y {
        return lhs.y < rhs.y
      }
      return lhs.x < rhs.x
    }
  }

  private struct BinaryHeap<T: Comparable> {
    private var data: [T] = []

    var isEmpty: Bool { data.isEmpty }

    mutating func push(_ item: T) {
      data.append(item)
      siftUp(from: data.count - 1)
    }

    mutating func pop() -> T? {
      guard !data.isEmpty else { return nil }

      if data.count == 1 {
        return data.removeLast()
      }

      let result = data[0]
      data[0] = data.removeLast()
      siftDown(from: 0)
      return result
    }

    private mutating func siftUp(from index: Int) {
      var child = index
      while child > 0 {
        let parent = (child - 1) >> 1
        guard data[child] < data[parent] else { break }
        data.swapAt(child, parent)
        child = parent
      }
    }

    private mutating func siftDown(from index: Int) {
      var parent = index
      let count = data.count

      while true {
        let left = (parent << 1) + 1
        let right = left + 1
        var candidate = parent

        if left < count && data[left] < data[candidate] {
          candidate = left
        }

        if right < count && data[right] < data[candidate] {
          candidate = right
        }

        if candidate == parent { break }

        data.swapAt(parent, candidate)
        parent = candidate
      }
    }
  }

  private func distance(ax: Int, ay: Int, bx: Int, by: Int) -> Int {
    return abs(ax - bx) + abs(ay - by)
  }

  private func packCoords(x: Int, y: Int) -> Int {
    return y * width + x
  }

  private func unpackCoords(packed: Int) -> (x: Int, y: Int) {
    return (packed % width, packed / width)
  }

  private func findPath() -> (path: [(x: Int, y: Int)]?, nodesExplored: Int) {
    let grid = mazeGrid

    let gScores = gScoresCache
    let cameFrom = cameFromCache

    let size = width * height
    gScoresCache.withUnsafeMutableBufferPointer { buffer in
      buffer.assign(repeating: Int.max)
    }
    cameFromCache.withUnsafeMutableBufferPointer { buffer in
      buffer.assign(repeating: -1)
    }

    var openSet = BinaryHeap<Node>()
    var nodesExplored = 0

    let startIdx = packCoords(x: startX, y: startY)
    gScoresCache[startIdx] = 0

    openSet.push(
      Node(
        x: startX, y: startY,
        fScore: distance(ax: startX, ay: startY, bx: goalX, by: goalY)))

    let directions = [(0, -1), (1, 0), (0, 1), (-1, 0)]

    while let current = openSet.pop() {
      nodesExplored += 1

      if current.x == goalX && current.y == goalY {

        var path: [(x: Int, y: Int)] = []
        var x = current.x
        var y = current.y

        while x != startX || y != startY {
          path.append((x, y))
          let idx = packCoords(x: x, y: y)
          let packed = cameFromCache[idx]
          if packed == -1 { break }

          let coords = unpackCoords(packed: packed)
          x = coords.x
          y = coords.y
        }

        path.append((startX, startY))
        path.reverse()
        return (path, nodesExplored)
      }

      let currentIdx = packCoords(x: current.x, y: current.y)
      let currentG = gScoresCache[currentIdx]

      for (dx, dy) in directions {
        let nx = current.x + dx
        let ny = current.y + dy

        if nx < 0 || nx >= width || ny < 0 || ny >= height { continue }
        if !grid[ny][nx] { continue }

        let tentativeG = currentG + 1000
        let neighborIdx = packCoords(x: nx, y: ny)

        if tentativeG < gScoresCache[neighborIdx] {

          cameFromCache[neighborIdx] = currentIdx
          gScoresCache[neighborIdx] = tentativeG

          let fScore = tentativeG + distance(ax: nx, ay: ny, bx: goalX, by: goalY)
          openSet.push(Node(x: nx, y: ny, fScore: fScore))
        }
      }
    }

    return (nil, nodesExplored)
  }

  private var gScoresCache: [Int]
  private var cameFromCache: [Int]

  private var resultVal: UInt32 = 0
  private var startX: Int = 1
  private var startY: Int = 1
  private var goalX: Int = 0
  private var goalY: Int = 0
  private var width: Int = 0
  private var height: Int = 0
  private var mazeGrid: [[Bool]] = []

  init() {

    gScoresCache = []
    cameFromCache = []
  }

  func prepare() {

    width = Int(configValue("w") ?? 50)
    height = Int(configValue("h") ?? 50)

    goalX = width - 2
    goalY = height - 2

    mazeGrid = MazeGenerator.Maze.generateWalkableMaze(width: width, height: height)

    let size = width * height
    gScoresCache = [Int](repeating: Int.max, count: size)
    cameFromCache = [Int](repeating: -1, count: size)
  }

  func run(iterationId: Int) {
    let (path, nodesExplored) = findPath()

    var localResult: UInt32 = 0

    let pathLength = UInt32(path?.count ?? 0)
    localResult = pathLength

    localResult = (localResult << 5) &+ UInt32(nodesExplored)

    resultVal &+= localResult
  }

  var checksum: UInt32 {
    return resultVal
  }

  func name() -> String {
    return "AStarPathfinder"
  }
}

File: julia/src/benchmarks/AStarPathfinder.jl
mutable struct AStarPathfinder <: AbstractBenchmark
    width::Int64
    height::Int64
    start_x::Int64
    start_y::Int64
    goal_x::Int64
    goal_y::Int64
    maze_grid::Vector{Vector{Bool}}
    g_scores_cache::Vector{Int64}
    came_from_cache::Vector{Int64}
    result::UInt32
end

struct Node
    x::Int64
    y::Int64
    f_score::Int64
end

function Base.isless(a::Node, b::Node)::Bool
    if a.f_score != b.f_score
        return a.f_score < b.f_score
    end
    if a.y != b.y
        return a.y < b.y
    end
    return a.x < b.x
end

mutable struct BinaryHeap{T}
    data::Vector{T}

    function BinaryHeap{T}() where {T}
        new{T}(Vector{T}())
    end
end

function heap_push!(heap::BinaryHeap{T}, item::T) where {T}
    Base.push!(heap.data, item)
    sift_up!(heap, length(heap.data))
end

function heap_pop!(heap::BinaryHeap{T})::Union{T,Nothing} where {T}
    if Base.isempty(heap.data)
        return nothing
    end
    if length(heap.data) == 1
        return Base.pop!(heap.data)
    end

    result = heap.data[1]
    heap.data[1] = Base.pop!(heap.data)
    sift_down!(heap, 1)
    return result
end

function heap_isempty(heap::BinaryHeap)::Bool
    return Base.isempty(heap.data)
end

function sift_up!(heap::BinaryHeap{T}, index::Int64) where {T}
    while index > 1
        parent = index ÷ 2
        if Base.isless(heap.data[index], heap.data[parent])
            heap.data[index], heap.data[parent] = heap.data[parent], heap.data[index]
            index = parent
        else
            break
        end
    end
end

function sift_down!(heap::BinaryHeap{T}, index::Int64) where {T}
    n = length(heap.data)
    while true
        left = 2 * index
        right = left + 1
        smallest = index

        if left <= n && Base.isless(heap.data[left], heap.data[smallest])
            smallest = left
        end
        if right <= n && Base.isless(heap.data[right], heap.data[smallest])
            smallest = right
        end

        if smallest == index
            break
        end

        heap.data[index], heap.data[smallest] = heap.data[smallest], heap.data[index]
        index = smallest
    end
end

function manhattan_distance(ax::Int64, ay::Int64, bx::Int64, by::Int64)::Int64
    return abs(ax - bx) + abs(ay - by)
end

function pack_coords(width::Int64, x::Int64, y::Int64)::Int64
    return (y - 1) * width + x
end

function unpack_coords(width::Int64, packed::Int64)::Tuple{Int64,Int64}
    packed -= 1
    y = packed ÷ width
    x = packed - y * width
    return (x + 1, y + 1)
end

function find_path(
    astar::AStarPathfinder,
)::Tuple{Union{Vector{Tuple{Int64,Int64}},Nothing},Int64}
    grid = astar.maze_grid
    width = astar.width
    height = astar.height

    size_val = width * height
    g_scores = astar.g_scores_cache
    came_from = astar.came_from_cache

    Base.fill!(g_scores, typemax(Int64))
    Base.fill!(came_from, -1)

    open_set = BinaryHeap{Node}()
    nodes_explored = 0

    start_idx = pack_coords(width, astar.start_x, astar.start_y)
    g_scores[start_idx] = 0

    start_f = manhattan_distance(astar.start_x, astar.start_y, astar.goal_x, astar.goal_y)
    heap_push!(open_set, Node(astar.start_x, astar.start_y, start_f))

    directions = [(0, -1), (1, 0), (0, 1), (-1, 0)]

    while !heap_isempty(open_set)
        current = heap_pop!(open_set)
        nodes_explored += 1

        if current.x == astar.goal_x && current.y == astar.goal_y

            path = Vector{Tuple{Int64,Int64}}()
            x = current.x
            y = current.y

            while x != astar.start_x || y != astar.start_y
                Base.push!(path, (x, y))
                idx = pack_coords(width, x, y)
                packed = came_from[idx]
                if packed == -1
                    break
                end
                x, y = unpack_coords(width, packed)
            end

            Base.push!(path, (astar.start_x, astar.start_y))
            Base.reverse!(path)
            return (path, nodes_explored)
        end

        current_idx = pack_coords(width, current.x, current.y)
        current_g = g_scores[current_idx]

        for (dx, dy) in directions
            nx = current.x + dx
            ny = current.y + dy

            if nx < 1 || nx > width || ny < 1 || ny > height
                continue
            end
            if !grid[ny][nx]
                continue
            end

            tentative_g = current_g + 1000
            neighbor_idx = pack_coords(width, nx, ny)

            if tentative_g < g_scores[neighbor_idx]
                came_from[neighbor_idx] = current_idx
                g_scores[neighbor_idx] = tentative_g

                f_score =
                    tentative_g + manhattan_distance(nx, ny, astar.goal_x, astar.goal_y)
                heap_push!(open_set, Node(nx, ny, f_score))
            end
        end
    end

    return (nothing, nodes_explored)
end

function AStarPathfinder()
    width = Helper.config_i64("AStarPathfinder", "w")
    height = Helper.config_i64("AStarPathfinder", "h")
    start_x = Int64(2)
    start_y = Int64(2)
    goal_x = width - 1
    goal_y = height - 1
    maze_grid = Vector{Vector{Bool}}()

    size_val = width * height
    g_scores_cache = Base.fill(typemax(Int64), size_val)
    came_from_cache = Base.fill(-1, size_val)

    return AStarPathfinder(
        width,
        height,
        start_x,
        start_y,
        goal_x,
        goal_y,
        maze_grid,
        g_scores_cache,
        came_from_cache,
        UInt32(0),
    )
end

name(b::AStarPathfinder)::String = "AStarPathfinder"

function prepare(b::AStarPathfinder)

    b.maze_grid = generate_walkable_maze(b.width, b.height)
end

function run(b::AStarPathfinder, iteration_id::Int64)
    path, nodes_explored = find_path(b)

    local_result = UInt32(0)

    path_length = path === nothing ? 0 : length(path)
    local_result = UInt32(path_length)

    local_result = (local_result << 5) + UInt32(nodes_explored)

    b.result += local_result
end

function checksum(b::AStarPathfinder)::UInt32
    return b.result
end
